### DISCUSSION TITLE
IsExpanded Setter issue

### DISCUSSION THREAD

**Discussion by cristinathoughtpennies** [AvaloniaUI/Avalonia#9719] [Q&A]
Hello! I am trying to either expand a TreeViewItem when it is selected or be able to set whether the currently selected TreeViewItem is expanded. This is what I have right now, but when the item is de-selected, it un-expands which is not what i want. Any help would be useful thanks!
```
<TreeView.Styles>
<Style Selector="TreeViewItem:selected">
	<Setter Property="IsExpanded" Value="True"/>
</Style>
</TreeView.Styles>
```

**Comment by maxkatz6** [comment]
IsExpanded is a direct property (or data property), which can't be styled. In newer versions it will throw an exception as well.

**Reply by cristinathoughtpennies** [reply]
Ah ok, what would be the proper way to programmatically expand a TreeViewItem then? I am using DataTemplates to fill out the TreeViewItems

**Reply by robloo** [reply]
I don't believe it is possible since it is a DirectProperty. You cannot set this value from a Style...

@maxkatz6 @grokys This fundamentally is why I think even maintainers are misusing direct properties. In my mind direct properties should be almost exclusively used for "content-type" properties and lists. "Data" is a good restriction too I suppose but IsExpanded certainly isn't "data" it's "state".

With that said, I'm taking advantage of IsExpanded being a direct property to support cancellation before visual state changes... to undo this would require an OnPropertyChanging method/event. Might be good to consider for 11.0 as this is a fundamental issue I think (something WPF also doesn't support I realize).

**Reply by cristinathoughtpennies** [reply]
Thanks for the reply! Glad I didn't spend more time on it if programmatically expanding a TreeViewItem is impossible to do haha

**Reply by cristinathoughtpennies** [reply]
Hello again! Turns out that being able to force TreeViewItems to expand is important to my project, ex: for functionality like being able to search and Go To an item in the tree.
Is there _anything_ that I can do to get this functionality into my project?

**Reply by timunie** [reply]
You can create a behavior (or code behind) for example where you loop over each tree view item and expand it. You may need to do it recursive.

**Reply by jp2masa** [reply]
You can use a converter too, like https://github.com/jp2masa/Movere/blob/f981f5da6f018ef8d730ed7b748268a0420d4368/src/Movere/Converters/FileExplorerTreeIsExpandedConverter.cs. You have to check if the selected node belongs to the target node's subtree, which in that converter was simplified in line 26 as `parent.FullPath.StartsWith(itemFolder.FullPath, StringComparison.Ordinal)`, but if you are working with a tree without paths you can just implement a `BelongsToSubtree` method for example.

**Reply by cristinathoughtpennies** [reply]
@timunie Thanks for the response! Is there a code example you could point me to of a recursive behavior or code behind?

**Reply by cristinathoughtpennies** [reply]
@jp2masa Thanks for the response! Im a bit confused, since in the comments above, @maxkatz6 and @robloo say that its impossible to set IsExpanded from a Style, but that's what I'm seeing in the code you linked, is there something Im missing?

**Reply by jp2masa** [reply]
I'm probably doing it wrong then, it worked by chance.

There are many other solutions possible:

- Behaviors, as @timunie suggested
- Subclass items control and create a custom generator which sets the property to the appropriate binding
- A custom attached property (which can be used in styles), with a property changed handler that sets the direct property

**Reply by timunie** [reply]
> @timunie Thanks for the response! Is there a code example you could point me to of a recursive behavior or code behind?

Sorry for the late reply. Here is the repo. It has some great samples in the code. Just clone it and gigg into the samples :-) https://github.com/wieslawsoltes/AvaloniaBehaviors
